kernel/sched: Use kernel timeouts for timeslice expirations

andyross opened the pull request:
[~~Not ready for merge yet, posted just for discussion.~~  This is how I think we want
to address the core bug in #55444, ~~please discuss there, not here~~.  (*Edit: this is working out and I think I'd like to merge this one.  Please review here, for real.*) It's quite a bit smaller
and more orthogonally leverages existing kernel features (timeouts) instead of perpetuating
the microdisaster with doing the timeslice accounting separately.  ~~It also doesn't quite work yet,
the measured times in the scheduler_api test are off by a tick and I'm trying to figure out if I've
simply messed something up or if that's a latent bug in the old version~~]

Rework the fragile and ad-hoc computation of timeslice expirations
into per-CPU struct _timeout objects with regular callbacks.  The
expiration callbacks themselves simply set a per-cpu flag (they might
run on any CPU), which gets checked at the end of the timer ISR on
every CPU.

This simplifies logic and removes a bunch of code.  It also fixes at
least three bugs:

1. On SMP, the number of ticks announced on any given CPU is going to
be a subset of all expired ticks.  This broke the accounting of
timeslice ticks, and effectively meant that timeslicing only worked on
SMP on systems where one CPU could hog all the announcements, and only
on that CPU.

2. The bootstrap path to set the arm the timer driver after setting
the first timeout couldn't take into account sys_clock_elapsed()
ticks, as it didn't know whether it was being called underneath an
existing announce loop.  Now this code is no longer responsible for
knowing anything about time slicing at all.

3. Also on SMP, there was a case where two CPUs timeslicing
simultaneously could stomp on each others' timeouts in
z_set_timeout_expiry(), as neither had a way of knowing what the
other's state was.  Now, timeouts are global objects with simple
expiration times, and there's no need for that function at all.

andyross committed:
subsys/pm: Use correct timeout API

The z_set_timeout_expiry() function is a wrapper provided in the
timeout subsystem that has some corrections for timelicing.  But the
suspend code doesn't care; by definition there are no active threads
that might be timeslicing at the point where the OS is suspending.
Use the lower-level timer driver API instead.

(z_set_timeout_expiry() is also about to disappear)

Signed-off-by: Andy Ross <andyross@google.com>

andyross committed:
kernel/sched: Use kernel timeouts for timeslice expirations

Rework the fragile and ad-hoc computation of timeslice expirations
into per-CPU struct _timeout objects with regular callbacks.  The
expiration callbacks themselves simply set a per-cpu flag (they might
run on any CPU), which gets checked at the end of the timer ISR on
every CPU.

This simplifies logic and removes a bunch of code.  It also fixes at
least three bugs:

1. As @npitre discovered: On SMP, the number of ticks announced on any
given CPU is going to be a subset of all expired ticks.  This broke
the accounting of timeslice ticks, and effectively meant that
timeslicing only worked on SMP on systems where one CPU could hog all
the announcements, and only on that CPU.

2. The bootstrap path to arm the timer driver after setting the first
timeout in an empty list couldn't take into account
sys_clock_elapsed() ticks, as it didn't know whether it was being
called underneath an existing announce loop.  Now this code is no
longer responsible for knowing anything about time slicing at all.

3. Also on SMP, there was a case where two CPUs timeslicing
simultaneously could stomp on each others' timeouts in
z_set_timeout_expiry(), as neither had a way of knowing what the
other's state was.  CPUs could miss their own expiration and have to
wait for the slice expiration on the other CPU.  Now, timeouts are
global objects with simple expiration times, and there's no need for
that function at all.

Signed-off-by: Andy Ross <andyross@google.com>

andyross commented on kernel/sched.c:
```diff
@@ -405,6 +405,8 @@ static void move_thread_to_end_of_prio_q(struct k_thread *thread)
 static int slice_ticks;
 static int slice_max_prio;
+struct _timeout slice_timeouts[CONFIG_MP_MAX_NUM_CPUS];
+bool slice_expired[CONFIG_MP_MAX_NUM_CPUS];
```
No particular reason these couldn't go into the _current_cpu struct, though "struct _timeout" has a header ordering issue that would need to be fixed.  They're here just because this is where we have the timeslicing data right now.

andyross replied:
Another future improvement might be to try to unify this struct _timeout with the one already in the thread struct.  That was a little hairy, as there are likely lots of places that assume the thread timeout will only ever be used for suspended threads.  But it should definitely be possible: needless to say a thread can never be simultaneously running and timeslicing and suspended.

andyross commented on kernel/sched.c:
```diff
@@ -501,15 +512,8 @@ static k_spinlock_key_t slice_expired_locked(k_spinlock_key_t sched_lock_key)
-	 * Also prevents any nested interrupts from changing
-	 * thread state to avoid similar issues, since this would
-	 * normally run with IRQs enabled.
-	 */
```
Sorta specious comment got removed.  Not strictly related, the lock is still present (for the obvious reasons mentioned)

andyross commented on kernel/sched.c:
```diff
@@ -522,19 +526,15 @@ void z_time_slice(int ticks)
-			_current_cpu->slice_ticks -= ticks;
 		}
-	} else {
-		_current_cpu->slice_ticks = 0;
```
Look at all the logic that goes away when the tick accounting happens just once at reset time

andyross commented on kernel/timeout.c:
```diff
@@ -77,11 +77,6 @@ static int32_t next_timeout(void)
-	if (_current_cpu->slice_ticks && _current_cpu->slice_ticks < ret) {
-		ret = _current_cpu->slice_ticks;
-	}
-#endif
```
Note how **all** the timeslice logic disappears from timeout.c now, with the sole exception of calling z_time_slice() out of the ISR.  Timeouts merely provide a generic API to the scheduler now, which is how is should have been.

andyross commented:
Thing I got it.  It was indeed on me: the new code left in a bit that ended up double-adding the sys_clock_elapsed() ticks.  So... I guess I can spin that as good news: this version removes yet more code.  Haven't done a full twister locally yet, so we'll see what CI says.  But I think I'm going to remove the DNM, this is ready for real review.  @npitre can you try this on your SMP rig and see if it addresses the problems?

npitre commented:
I don't see how this solves anything.

Scenario:

- CPU0 adds whatever timeout to expire at t+2. It is the first timeout
  in the list so the hardware timer on CPU0 is set to fire accordingly.

- CPU1 adds a timeslice timeout to expire at t+3. It is not the first
  timeout so it is simply queued.

- The timer IRQ is triggered on CPU0 at t+2. The corresponding timeout is
  processed and the hardware timer re-armed for the next timeout at t+3.

- The timer IRQ is triggered at t+3, again on CPU0, as this is the CPU
  that rearmed the timer previously. It calls `slice_timeout()` which sets 
  `slice_expired[0]` to true.

- CPU0 completes its execution of `sys_clock_announce()` by calling
  `z_time_slice()`. Because `slice_timeout()` is true the thread on CPU0
  is timesliced away while thread on CPU1 is unaffected.

npitre commented:
> Look at all the logic that goes away when the tick accounting happens just once at reset time

Indeed. I got rid of that part as well in my version.

andyross commented:
Timer interrupts are required to happen on all cpus, is that the part that's mixing you up?  Does this really not work with your test that was failing on the SMP hardware?    Can you cook something up I can run locally with an emulator platform?  I feel really quite strongly that this is the right tactic here, and not the computation of a global timeout on each CPU individually.

npitre commented:
> Timer interrupts are required to happen on all cpus, is that the part that's
> mixing you up?  Does this really not work with your test that was failing on
> the SMP hardware?

No can do. Both the ARM Architected Timer and the RISC-V machine timer are
per-CPU timers. They all use a global counter, but the match registers are per
CPU. I'm sure there are more like that out there.

Furthermore, the doc says:

* Zephyr is agnostic about which CPU services timer interrupts. It is not
  illegal (though probably undesirable in some circumstances) to have
  every timer interrupt handled on a single processor.

It is far easier to assume per_CPU timers and cope with those architectures
whose timer interrupts all CPUs than trying to have per-CPU timer architectures
to create synchronous timer interrupts on all CPUs.

Therefore here is no way your approach can work without directed IPIs.

First approach:

- Use generic global timeouts to register timeslice expiry, one timeout
  struct per CPU. Put those structs in array so that you can retrieve
  the related CPU number from the struct position.

- Upon timeout expiration, determine target CPU per above. If this is
  the current CPU then call `z_time_slice()`. If not the current CPU then
  use a directed IPI to have `z_time_slice()` executed on the target CPU.

This is basically your approach. Directed IPIs are already implemented on 
ARM64 and RISC-V to deal with FPU context retrieval from another CPU.

Alternative approach:

- Have the notion of global timeouts that can be processed on any CPU (what we
  already have) and local timeouts that are tied to individual CPUs.

- When a timeslice sets its expiration, it uses a local timeout.
  When that local timeout expires, it is already on the right CPU.
  Suffice to call `z_time_slice()` directly.

This is my approach. I isolated that part in RP #55498. The current code is
more or less already geared for this. Tracking local timeouts is rather simple, especially if we limit ourselves to
only one per CPU. No extra IPI is needed, meaning this works even if a 
particular target can't do targetted IPIs, or any kind of IPIs.

I personally lean towards the second approach simply because it is more
run time efficient and doesn't require any extra arch specific support.

andyross commented:
Ah... OK, I see a hole, which might be what you were pointing at.  It's not enough to just z_time_slice() at the end of the local timer interrupt, we need to be sure all the ticks have been processed on the other CPUs too.  That requires a loop, though, something to check for a quiescent state (e.g. an empty queue, a flag set, something like that).  And I don't see where either one of us has that?  Your version does a single check at the end, which is more than mine, but AFAICT still not enough.

Will cook something up for this tomorrow morning.  FWIW: this is exactly the situation (receiving a timeslice expiration for a non-local CPU) where a targetted IPI would be perfect, which is what I was talking about earlier.  I fooled myself into thinking it wasn't needed in the short term, but there will need to be a work around in place until it's ready.

andyross commented:
> No can do. Both the ARM Architected Timer and the RISC-V machine timer are
per-CPU timers.

OK, well that's your problem then. :)  Yeah, that's not the earlier architecture..  If you dig, there's absolutely a discussion in the docs about how timeslicing interacts with the timer interrupt, I remember writing it.  Now I'm understanding why things looked wrong to you.  In any case, you're likely going to have serious problems with your patch on x86_64 and xtensa if you make the assumption that the local call to sys_clock_set_timeout() results in an interrupt on the local CPU.  That's just not the case.

Nonetheless, I'm sure we can make this work.  But I'm getting firmer and firmer here; it absolutely has to be made to work with "timeslice expiration as timeouts".  I suspect I can have something for you that works with "timeouts on any CPU".

andyross commented:
> Therefore here is no way your approach can work without directed IPIs.

Famous last words.  :)

npitre commented:
> Can you cook something up I can run locally with an emulator platform?

Sure!  

- Check out PR #55444, or the timeslice2 branch from my Zephyr fork at
  https://github.com/npitre/zephyr/.

- Do `west build -t run -b qemu_cortex_a53_smp tests/kernel/sched/timeslicing/`

npitre commented:
> OK, well that's your problem then. :)  Yeah, that's not the earlier
> architecture..  If you dig, there's absolutely a discussion in the 
> docs about how timeslicing interacts with the timer interrupt, I 
> remember writing it.  Now I'm understanding why things looked wrong 
> to you.  In any case, you're likely going to have serious problems 
> with your patch on x86_64 and xtensa if you make the assumption that 
> the local call to sys_clock_set_timeout() results in an interrupt on 
> the local CPU.  That's just not the case.

Andy... please.

Instead of trying to reimplement everything in the most different
way possible, I'd invite you to have a closer look at what I did for
a moment.

For the record I _did_ modify the hpet driver to make my super slicing test
work with my approach on x86. I'm quoting those changes here:

```
--- a/drivers/timer/hpet.c
+++ b/drivers/timer/hpet.c
@@ -272,6 +272,48 @@ static inline void hpet_timer_comparator_set_safe(uint64_t
next)
        }
 }

+#if defined(CONFIG_SMP) && defined(CONFIG_TICKLESS_KERNEL) && defined(CONFIG_TIMESLICING)
+/*
+ * We need per-CPU timeouts for timeslicing support.
+ *
+ * Timeouts are usually set per CPU. The HPET being a single system-wide
+ * device, we have to track per-CPU requests by hand and set the soonest
+ * non-expired timeout among them.
+ */
+static __pinned_bss uint64_t timeout_cpu[CONFIG_MP_MAX_NUM_CPUS];
+static __pinned_bss uint64_t last_timeout;
+
+__pinned_func
+static void process_percpu_timeouts(uint64_t now)
+{
+       uint64_t next_timeout = UINT64_MAX;
+
+       for (int i = 0; i < arch_num_cpus(); i++) {
+               if (timeout_cpu[i] <= now) {
+                       /* this one is expired */
+                       timeout_cpu[i] = 0;
+               } else if (timeout_cpu[i] < next_timeout) {
+                       /* this one is the best next candidate so far */
+                       next_timeout = timeout_cpu[i];
+               }
+       }
+       if (next_timeout != UINT64_MAX && next_timeout != last_timeout) {
+               last_timeout = next_timeout;
+               hpet_timer_comparator_set_safe(next_timeout);
+       }
+}
+
+static inline void set_this_cpu_timeout(uint64_t val)
+{
+       timeout_cpu[_current_cpu->id] = val;
+       process_percpu_timeouts(0);
+}
+
+#else
+#define process_percpu_timeouts(val) /* nothing */
+#define set_this_cpu_timeout(val) hpet_timer_comparator_set_safe(val)
+#endif
+
 __isr
 static void hpet_isr(const void *arg)
 {
@@ -315,6 +357,7 @@ static void hpet_isr(const void *arg)
                hpet_timer_comparator_set_safe(next);
        }

+       process_percpu_timeouts(last_count);
        k_spin_unlock(&lock, key);
        sys_clock_announce(dticks);
 }
@@ -369,7 +412,7 @@ void sys_clock_set_timeout(int32_t ticks, bool idle)
        k_spinlock_key_t key = k_spin_lock(&lock);
        uint64_t cyc = (last_tick + last_elapsed + ticks) * cyc_per_tick;

-       hpet_timer_comparator_set_safe(cyc);
+       set_this_cpu_timeout(cyc);
        k_spin_unlock(&lock, key);
 #endif
 }
```

Hardly a big deal.

npitre commented:
And if you prefer x86 to run my test:

- Check out PR #55444, or the timeslice2 branch from my Zephyr fork at
  htps://github.com/npitre/zephyr/.

- Do `west build -t run -b qemu_x86_64 tests/kernel/sched/timeslicing/`

tmleman commented:
@marcinszkudlinski fyi

andyross committed:
kernel/sched: Close hole with cross-core timeslice expirations

Moving timeslice events to timeouts isn't quite enough on SMP, as it's
still possible for systems that don't broadcast their timer interrupts
to end up handling an expiration for a foreign CPU.  There, we need an
IPI, and a symmetric call to z_time_slice() (which is itempotent and
fast) in the IPI ISR.

Signed-off-by: Andy Ross <andyross@google.com>

andyross commented:
Update with the simplest possible IPI integration.  We can update trivially to a core-targetted IPI in the future, and platforms that have timer drivers with the capability of registering separate per-cpu interrupts can just stub this out.

And @npitre: I promise I'm not trying to do this deliberately differently.  I'm trying to do it more simply (I mean, just look at the line count of the PR: this is touching fewer than 100 lines and removing 2x as many as it adds).  The scheme you have doesn't seem to understand the case of hardware like cavs_timer or HPET that can only deliver global interrupts, so it leaves existing bugs in place.  AFAICT, this works everywhere.

andyross commented:
To be clearer: let's get an existing solution in place that works with the existing timer drivers as-is.  If there are obvious improvements to be made at that point, let's do them.

andyross commented:
We should take a day or so to cool off, maybe.  I promise you there's no disrespect intended.  This is just an area that has been a thorn for a long time and a solution that's been bouncing around in my head for several years.

npitre commented:
> We should take a day or so to cool off, maybe.  I promise you there's no
> disrespect intended.

OK good.

> This is just an area that has been a thorn for a long time and a solution 
> that's been bouncing around in my head for several years.

Is that, perhaps, the reason why you had this knee jerk reaction when
presented with the possibility that my solution could actually work?

Granted, it wasn't "easy". Took me a while to figure it out too.

Now... I'm torn with regard to your solution. I do like its simplicity,
but I'm less pleased by its runtime IPI ping-pong overhead. If CPU0
gets the IRQ for CPU1's slice timeout, CPU0 will raise an IPI for CPU1
to notice. CPU1 will then modifi the run queue and, in turn, raise another IPI for
others to notice.

But hey, with my other scheduler patches on top, I can confirm that my
big SMP slicing test does pass. Therefore I therefore bow before you and
approve this PR.

andyross commented:
> CPU1 will then modifi the run queue and, in turn, raise another IPI for others to notice.

That's been a timeslicing misfeature for a long time, actually.  At some point I remember playing with code that would be able to recognize the noop nature of a timeslice reset (the common case is that you don't actually get preempted) and avoid the IPI, but something wasn't going to work easily and I gave up; I forget the details.

But as mentioned elsewhere, on hardware that does have the capability to deliver a timer interrupt specific to a CPU, it wouldn't be a huge change to introduce a second per-CPU timeout queue and an API to go with it (and one in the timer driver, obviously).  Then the announce loop could just zip them together merge-sort-style with not too much complexity.  On such a system, we'd never hit the IPI case (and could probably #if it out if there aren't any edge cases I'm missing).

andyross commented:
@peter-mitsis @teburd It would be god to do a full twister run on intel_adsp/ace before this merges, which is the other big SMP hardware target.   You'd think I'd be set up to do that for myself, but I'm actually not.

peter-mitsis reviewed (approved):
Nice clean implementation. :)

One nitpick (not something that would be a blocker)--naming. Can some of the time slice related symbols get a more distinct name? My poor brain was getting confused between keeping track of the differences between slice_timeout, slice_time, z_time_slice, slice_timeouts, ....  For example, even renaming slice_timeout() to slice_timeout_handler() helps my poor brain immensely.

(Aside: Due to computer issues, I am currently unable to run this on an xtensa platform)

teburd commented:
Please DNM until I've finished running tests on intel adsp platforms, will remove once done

teburd commented:
This passes all the kernel tests on cavs25 with gcc so unblocking my DNM
